collectors/ebpf.plugin/ebpf_socket.c

```diff
@@ -0,0 +1,381 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include <sys/resource.h>
+
+#include "ebpf.h"
+#include "ebpf_socket.h"
+
+/*****************************************************************
+ *
+ *  GLOBAL VARIABLES
+ *
+ *****************************************************************/
+
+static ebpf_functions_t socket_functions;
+
+static netdata_idx_t *socket_hash_values = NULL;
+static netdata_syscall_stat_t *socket_aggregated_data = NULL;
+static netdata_publish_syscall_t *socket_publish_aggregated = NULL;
+
+static char *socket_dimension_names[NETDATA_MAX_SOCKET_VECTOR] = { "sent", "received", "close", "sent", "received" };
+static char *socket_id_names[NETDATA_MAX_SOCKET_VECTOR] = { "tcp_sendmsg", "tcp_cleanup_rbuf", "tcp_close", "udp_sendmsg",
+                                                     "udp_recvmsg" };
+
+#ifndef STATIC
+/**
+ * Pointers used when collector is dynamically linked
+ */
+
+//Libbpf (It is necessary to have at least kernel 4.10)
+static int (*bpf_map_lookup_elem)(int, const void *, void *);
+static int (*bpf_map_delete_elem)(int fd, const void *key);
+
+static int *map_fd = NULL;
+/**
+ * End of the pointers
+ */
+#endif
+
+/*****************************************************************
+ *
+ *  PROCESS DATA AND SEND TO NETDATA
+ *
+ *****************************************************************/
+
+/**
+ * Update publish structure before to send data to Netdata.
+ *
+ * @param publish  the first output structure with independent dimensions
+ * @param tcp      structure to store IO from tcp sockets
+ * @param udp      structure to store IO from udp sockets
+ * @param input    the structure with the input data.
+ */
+static void ebpf_update_publish(netdata_publish_syscall_t *publish,
+                                netdata_publish_vfs_common_t *tcp,
+                                netdata_publish_vfs_common_t *udp,
+                                netdata_syscall_stat_t *input) {
+
+    netdata_publish_syscall_t *move = publish;
+    while(move) {
```

Comment:
What if `input` is NULL? Shouldn't we check it here? The same question for `ebpf_process.c`.

Reply:
The input is allocated with `callocz`, so case we cannot allocate it Netdata will shutdown the collector.

Reply:
Ok, but `input = input->next`. How you guarantee that you have allocated the same amount of elements as in `move` and put them in a linked list correctly?

Reply:
`input` and `publish` must be allocated with the same size always,.
I would not like to add this check here, because we will have a check every time that you need to send message to Netdata. I will bring a change that will ensure that when allocated both variables have the same size.
Another point that I would like to call attention, it is for the fact that these functions, like the function that does the allocation of the vectors are `static` inside each collector.